### ISSUE TITLE
Angle Degrees

### ISSUE THREAD

**Issue by eledahl** [TemugeB/joint_angles_calculate#1]
Hi @TemugeB! Read your post and it has helped me a ton. Just a quick question. How would you convert the angles to degrees? As in I am trying to measure starting ankle angle and then lowest angle in degrees throughout a squat but having difficulty getting the correct numbers.

**Comment by TemugeB** [comment]
You would simply call np.degrees() to convert radians to degrees. But I suspect your issue is not to do with conversion. If what you want is the smallest angle between the feet and the shins(see the image below), then you don't need to bother with any of my code. In 2D, the angle can be obtained by:
![image](https://user-images.githubusercontent.com/36071915/145533848-a0273c4f-e8ad-4fd6-9636-9f0e25e75de7.png)
Where A_hat and B_hat are unit vectors. 

If you want to calculate the 3D rotation angles, then you first have to define a coordinate system near the ankle and then express A and B vectors with respect to this coordinate system. This step is not necessary in 2D because you can express the A and B vectors in any coordinate system to get the smallest angle between them. However, in 3D, you must make sure that the initial coordinate system is properly defined and it also moves and rotates as the body also moves. If you've done this, then you just need to call get_R(A,B) function in utils.py to get the full rotation matrix. After that, call Decompose_R_ZXY() to obtain the rotation angles. Hope this helps.

![image](https://user-images.githubusercontent.com/36071915/145533962-c5b7efc0-e12a-400b-b24b-8b3f24d74384.png)

**Comment by eledahl** [comment]
Thank you @TemugeB! That is exactly what I am looking for. Appreciate the detailed answer!

**Comment by eledahl** [comment]
Hey @TemugeB, I've been trying to figure this out without having to bug you again but I am stuck with getting the correct angles still. 

Here's a sample code:

```
knee = [landmarks[mpPose.PoseLandmark.LEFT_KNEE.value].x,
                landmarks[mpPose.PoseLandmark.LEFT_KNEE.value].y, landmarks[mpPose.PoseLandmark.LEFT_KNEE.value].z]

ankle = [landmarks[mpPose.PoseLandmark.LEFT_ANKLE.value].x,
                 landmarks[mpPose.PoseLandmark.LEFT_ANKLE.value].y, landmarks[mpPose.PoseLandmark.LEFT_ANKLE.value].z]

toe = [landmarks[mpPose.PoseLandmark.LEFT_FOOT_INDEX.value].x,
               landmarks[mpPose.PoseLandmark.LEFT_FOOT_INDEX.value].y, landmarks[mpPose.PoseLandmark.LEFT_FOOT_INDEX.value].z]

def Get_R(A, B, C):
    BA = [A[0] - B[0], A[1] - B[1], A[2] - B[2]]
    BC = [C[0] - B[0], C[1] - B[1], C[2] - B[2]]

    # get unit vectors
    uA = BA/np.sqrt(np.sum(np.square(BA)))
    uB = BC/np.sqrt(np.sum(np.square(BC)))

    v = np.cross(uA, uB)
    s = np.sqrt(np.sum(np.square(v)))
    c = np.sum(uA * uB)

    vx = np.array([[0, -v[2], v[1]],
                   [v[2], 0, -v[0]],
                   [-v[1], v[0], 0]])

    R = np.eye(3) + vx + vx@vx*((1-c)/s**2)
    # print(R)
    thetaz = np.arctan2(-R[0, 1], R[1, 1])
    thetay = np.arctan2(-R[2, 0], R[2, 2])
    thetax = np.arctan2(R[2, 1], np.sqrt(R[2, 0]**2 + R[2, 2]**2))

    return np.degrees(thetaz), np.degrees(thetay), np.degrees(thetax)

print(Get_R(knee, ankle, toe))
```

Output through a squat:
```
(-166.16579222416192, -132.09030352095206, -68.98454307406247)
(-168.01811237020766, -133.77174680306527, -70.74135752884763)
(-175.54555951441964, -145.701864408284, -69.92086495204742)
(-175.48833265448116, -145.67027271651688, -69.41285286029752)
(-179.69254682465618, -151.23468594371454, -68.94372916172773)
(175.0868000856085, -158.00420518074458, -68.74187071269762)
(173.880435414026, -157.6431266327595, -68.82544763077827)
(170.4771257931608, -161.16472245230693, -69.7472013294359)
(155.21435665460172, 179.24715854369722, -68.02305948278112)
(147.07753832290476, 167.3820369041634, -65.56933647923016)
(143.3264462974199, 162.62702625332503, -64.9536123825029)
(143.7948022879432, 161.65024761148257, -65.26165456290454)
(142.65586720473723, 159.7541376715155, -65.24017212631617)
(143.2438654792192, 163.0599406954642, -65.78708237616917)
(143.4766096491628, 165.03416609951648, -66.10910509719494)
(143.46272014798492, 165.8441089582439, -66.03448795055013)
(143.8531681324156, 166.73096809827442, -66.01052775755191)
(144.43025500885497, 167.66550613512453, -65.9410577639059)
(144.76657047030227, 168.26614889629053, -66.02265029547384)
(144.97912503720116, 168.5416429769771, -65.99358165625215)
(144.98698053156053, 168.61883361857497, -65.73957952035713)
(145.36910890198976, 169.32044201678613, -65.31726905718192)
(145.22757836341128, 169.1329166646608, -65.14355605912498)
(145.58524189403036, 169.5932742869789, -65.11480663021224)
(145.84592404165434, 169.9557269859522, -65.15248623337231)
(147.27520853915178, 171.76333795239657, -65.01757947315137)
(148.54048065929473, 172.4905049237727, -65.10180547013114)
(148.6186855323799, 172.39421036287663, -64.98679962186223)
(148.37957904539456, 172.06255099291502, -65.09791801722888)
(148.20807326446388, 171.44935059531042, -65.02706128457032)
(147.5739981703111, 170.42671688544127, -65.10841933547053)
(147.88231961391818, 170.94101807215313, -64.88259191350872)
(148.05125316839957, 171.16479897149327, -64.7723335721334)
(148.12311912757954, 171.60704932279103, -64.68456325245653)
(148.72743290763992, 173.20142987852915, -64.77313373935337)
(149.25926324390738, 174.05946774439994, -65.1319128027858)
(149.974880575414, 175.64878374064278, -65.27348199545307)
(150.91637782798264, 177.1911464514869, -65.26003254201794)
(151.48199577098381, 178.95260215860577, -65.55599323606856)
(152.1544549095824, 179.858717099781, -65.22556697360655)
(155.28248860147642, -176.30423078021647, -65.40408504801285)
(155.64785777644732, -176.95645709724604, -65.9953520440599)
(155.56243924114364, -173.43800324856974, -67.68972942816609)
(155.28036558551642, -169.7170054775789, -69.47222124674036)
(157.6602699405618, -167.04070772143854, -69.99992520105567)
(156.50731004996828, -168.06917394184333, -69.75351860275373)
(158.3736254649831, -166.11637469764, -69.8818132657796)
(157.66431507205021, -166.1549741081191, -70.6576614239808)
(160.85172627094823, -162.61043080282616, -70.81991696197767)
(167.49714941235567, -154.0951112205509, -70.23568528490927)
(169.59758921429446, -152.00443559301524, -70.43994501438357)
(170.16660245817178, -151.59517928204133, -70.5958800167181)
(171.75404459975152, -150.34634087917803, -70.99363853149087)
(178.22711319464455, -144.45260805189406, -71.19758009924493)
(-176.33768990066895, -138.7883137826047, -70.6548282204153)
(-174.99803675675452, -138.39656051867286, -71.76703581148693)
(-160.92601328329118, -126.67550594901991, -69.7744743060199)
(-156.32155225442258, -124.64468555851488, -68.11777149557196)
(-151.19779786351518, -122.1812125947644, -68.3368649636906)
(-148.99505856371835, -120.33694731886568, -68.3541387480372)
(-148.9742077544299, -119.67040937236051, -67.67485178314782)
(-150.40159276450782, -123.69690956434091, -67.90938082156218)
(-153.0352777499472, -129.5573771456139, -66.72542866904236)
(-153.45941166206168, -133.46288770391862, -66.62592521280577)
(-155.49524228001619, -135.40451779493668, -66.2090000804632)
(-153.94884818022055, -136.28310883440886, -66.62817767701239)
(-154.32981299329617, -136.1852230649996, -65.29120126611093)
(-154.13993007838968, -135.5510560570515, -64.9944488574347)
(-153.32064421642005, -134.65119377722937, -65.04913158016804)
(-153.5896394197292, -134.8873252447299, -64.68023712005463)
(-153.7285492485245, -134.48265679676857, -63.56020512585134)
(-155.3773054382042, -136.52348025990005, -64.17832630799138)
(-157.3431149075851, -136.75685525542954, -61.97245453434439)
(-157.1124350712844, -137.20353100360992, -62.71114448187762)
(-156.372618849337, -135.0871088742926, -62.3591729385569)
(-156.19661301339798, -134.3732990812593, -61.441503256991666)
(-154.40263546391014, -129.4989684869293, -60.78596241340617)
(-154.2109202611853, -129.3786777407406, -59.89589824109637)
(-154.37363490984228, -128.59747687477574, -59.18897635994057)
(-154.9349334992528, -130.14494269912362, -60.45631594970191)
(-155.34377570481917, -130.6468433794787, -60.29356136449488)
(-155.31114177121668, -130.81655665040736, -60.19148291489993)
(-154.48506341875063, -128.86607072480516, -59.85023566648196)
(-154.96517802336504, -128.95802128483817, -60.65771193852436)
(-156.50327639270833, -129.4807955541088, -60.23526949747725)
(-156.67093554565585, -127.26594261833986, -60.66423065547217)
(-157.28763517148454, -128.7101487819022, -59.974257794022215)
(-157.09140193199843, -128.43441247359493, -59.67817917933919)
(-173.09469165243502, -148.18972895081367, -59.91431543714544)
(-172.09189060285703, -146.72897109858064, -60.40434533733927)
(-163.7743142163868, -137.40233258575483, -59.61818520512998)
...
```
I'm guessing it is how I am defining the vectors or the conversion at the end?

**Comment by TemugeB** [comment]
I don't understand the question. Is it the jumping angle values? An angle of -179 and 179 result in nearly identical body poses but the angle values are exact opposites. The return value of ```np.arctan2``` is between [-pi, pi], which is why you are getting this result. Easiest way to fix this is to force [0, 2pi] values by adding 360 where necessary. But in general, the angles that make a rotation matrix is not unique and can cause discontinuity in the obtained values. This makes them not very good inputs to machine learning models. 
If you want to check for discontinuity, you can plot each angle with respect to time. It should form a continuous curve. Let me know if this answers your question.

**Comment by eledahl** [comment]
I may be using this incorrectly then. The result I am looking for would be to expect one of the starting angles of the ankle to be somewhere between 90 and let's say 110. As the person squats,then the angle would drop to between 60 and 80.

**Comment by TemugeB** [comment]
Are you trying to calculate angle directly from a single camera body pose estimate? I'm not sure if this is doable because the 3D coordinates returned from a single camera pose estimate does not have a well defined coordinate system, which can lead to problems that you are facing now. More safe way to do this would be to use two cameras and triangulate the points, which forces a specific worldspace coordinate system into the 3D coordinates.
In any case, I suggest you first try to calculate the 1D angle between shins and feet to make sure what you said above holds. 
In other words, something along this line
```python
theta =  np.arccos(np.dot(A, B))
```
where A and B are unit vectors representing the directions of the shins and feet.

**Comment by eledahl** [comment]
Welp that might have done it! With BlazePose's `pose_world_landmarks` and `theta =  np.arccos(np.dot(A, B))` I am getting angles in the 110s at top of squat down to the 80s at bottom of squat. Tested on two videos and will test more tomorrow. Thank you again for all of your help!
